From 6f39aea3a1a2a8b0baf177029c1844370986a4ff Mon Sep 17 00:00:00 2001
From: Jerry Chang <jerchang@nvidia.com>
Date: Thu, 18 Jan 2024 14:59:33 +0800
Subject: [PATCH] capture-ivc: fix multi-cam race condition

add semaphore to avoid multi-cam race condition

Bug 4425972

Change-Id: Idf318465ed5cf8fe16a5d395452ba104a9f7cbd3
Signed-off-by: Jerry Chang <jerchang@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-nvidia/+/3060234
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: Anubhav Rai <arai@nvidia.com>
Reviewed-by: Bibek Basu <bbasu@nvidia.com>
GVS: Gerrit_Virtual_Submit <buildbot_gerritrpt@nvidia.com>
---
 .../platform/tegra/rtcpu/capture-ivc-priv.h   |  9 +++++---
 drivers/platform/tegra/rtcpu/capture-ivc.c    | 21 ++++++++++++++++---
 2 files changed, 24 insertions(+), 6 deletions(-)

diff --git a/drivers/platform/tegra/rtcpu/capture-ivc-priv.h b/drivers/platform/tegra/rtcpu/capture-ivc-priv.h
index 04c95aef3..c4eaa4eec 100644
--- a/drivers/platform/tegra/rtcpu/capture-ivc-priv.h
+++ b/drivers/platform/tegra/rtcpu/capture-ivc-priv.h
@@ -1,9 +1,11 @@
-/**
+/* SPDX-License-Identifier: LicenseRef-NvidiaProprietary
+ *
+ * SPDX-FileCopyrightText: Copyright (c) 2017-2024 NVIDIA CORPORATION.
+ *                         All rights reserved.
+ *
  * @file drivers/platform/tegra/rtcpu/capture-ivc-priv.h
  * @brief Capture IVC driver private header for T186/T194
  *
- * Copyright (c) 2017-2019 NVIDIA Corporation.  All rights reserved.
- *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms and conditions of the GNU General Public License,
  * version 2, as published by the Free Software Foundation.
@@ -37,6 +39,7 @@ struct tegra_capture_ivc_cb_ctx {
 	tegra_capture_ivc_cb_func cb_func;
 	/** Private context of a VI/ISP capture context */
 	const void *priv_context;
+	struct semaphore sem_ch;
 };
 
 /**
diff --git a/drivers/platform/tegra/rtcpu/capture-ivc.c b/drivers/platform/tegra/rtcpu/capture-ivc.c
index c5e1bc519..48acf32fa 100644
--- a/drivers/platform/tegra/rtcpu/capture-ivc.c
+++ b/drivers/platform/tegra/rtcpu/capture-ivc.c
@@ -1,9 +1,11 @@
+// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
 /*
+ * SPDX-FileCopyrightText: Copyright (c) 2017-2024 NVIDIA CORPORATION.
+ *                         All rights reserved.
+ *
  * @file drivers/platform/tegra/rtcpu/capture-ivc.c
  * @brief Capture IVC driver
  *
- * Copyright (c) 2017-2022 NVIDIA Corporation.  All rights reserved.
- *
  * This program is free software; you can redistribute it and/or modify it
  * under the terms and conditions of the GNU General Public License,
  * version 2, as published by the Free Software Foundation.
@@ -28,11 +30,15 @@
 #include <linux/tegra-ivc.h>
 #include <linux/tegra-ivc-bus.h>
 #include <linux/nospec.h>
-
+#include <linux/semaphore.h>
 #include <asm/barrier.h>
 
 #include "capture-ivc-priv.h"
 
+/* Timeout for acquiring channel-id */
+#define TIMEOUT_ACQUIRE_CHANNEL_ID 120
+
+
 static int tegra_capture_ivc_tx(struct tegra_capture_ivc *civc,
 				const void *req, size_t len)
 {
@@ -165,6 +171,11 @@ int tegra_capture_ivc_notify_chan_id(uint32_t chan_id, uint32_t trans_id)
 
 	civc = __scivc_control;
 
+	if (down_timeout(&civc->cb_ctx[chan_id].sem_ch,
+				TIMEOUT_ACQUIRE_CHANNEL_ID)) {
+		return -EBUSY;
+	}
+
 	mutex_lock(&civc->cb_ctx_lock);
 
 	if (WARN(civc->cb_ctx[trans_id].cb_func == NULL,
@@ -269,6 +280,7 @@ int tegra_capture_ivc_unregister_control_cb(uint32_t id)
 	civc->cb_ctx[id].priv_context = NULL;
 
 	mutex_unlock(&civc->cb_ctx_lock);
+	up(&civc->cb_ctx[id].sem_ch);
 
 	/*
 	 * If it's trans_id, client encountered an error before or during
@@ -415,6 +427,9 @@ static int tegra_capture_ivc_probe(struct tegra_ivc_channel *chan)
 	mutex_init(&civc->cb_ctx_lock);
 	mutex_init(&civc->ivc_wr_lock);
 
+	for (i = 0; i < TOTAL_CHANNELS; i++)
+		sema_init(&civc->cb_ctx[i].sem_ch, 1);
+
 	/* Initialize ivc_work */
 	INIT_WORK(&civc->work, tegra_capture_ivc_worker);
 
-- 
2.25.1

